### PR TITLE
Support for netavark backed podman(4.x+)  - fixing #2882

### PR DESCRIPTION
**Pull request by wherka-ama**
Adding support for netavark backed podman(4.x+) 

See: https://github.com/kubernetes-sigs/kind/issues/2882

### CODE REVIEW

File: pkg/cluster/internal/providers/podman/provision.go [hunk 1]
```diff
@@ -251,28 +252,67 @@ func getProxyEnv(cfg *config.Cluster, networkName string) (map[string]string, er
...
-		noProxyList = append(noProxyList, ".svc", ".svc.cluster", ".svc.cluster.local")
+		noProxyList = append(noProxyList, ".svc", ".svc.cluster", ".svc.cluster.local", strings.Join([]string{cfg.Name, "control-plane"}, "-"))
 		noProxyJoined := strings.Join(noProxyList, ",")
 		envs[common.NOProxy] = noProxyJoined
 		envs[strings.ToLower(common.NOProxy)] = noProxyJoined
 	}
 	return envs, nil
 }
 
+type PodmanNetworks []struct {
```
Comment (BenTheElder) [inline pkg/cluster/internal/providers/podman/provision.go:268]: This type shouldn’t be exported.

File: pkg/cluster/internal/providers/podman/provision.go [hunk 2]
```diff
@@ -251,28 +252,67 @@ func getProxyEnv(cfg *config.Cluster, networkName string) (map[string]string, er
...
 		}
 		noProxyList := append(subnets, envs[common.NOProxy])
-		// Add pod and service dns names to no_proxy to allow in cluster
+		// Add pod, service and control plane(API server) dns names to no_proxy to allow in cluster
 		// Note: this is best effort based on the default CoreDNS spec
 		// https://github.com/kubernetes/dns/blob/master/docs/specification.md
 		// Any user created pod/service hostnames, namespaces, custom DNS services
 		// are expected to be no-proxied by the user explicitly.
-		noProxyList = append(noProxyList, ".svc", ".svc.cluster", ".svc.cluster.local")
+		noProxyList = append(noProxyList, ".svc", ".svc.cluster", ".svc.cluster.local", strings.Join([]string{cfg.Name, "control-plane"}, "-"))
```
Comment (BenTheElder) [inline pkg/cluster/internal/providers/podman/provision.go:260]: Changes to this method seem unrelated to netavark support and should also be kept consistent with the docker implementation

**Review by BenTheElder — COMMENTED** [review]
Thanks for the PR!

File: pkg/cluster/internal/providers/podman/provision.go (hunk 2 above)
Comment (BenTheElder) [reply pkg/cluster/internal/providers/podman/provision.go:260]: This should be a separate commit or PR, and it should probably not reimplement naming the control plane endpoint with different logic.

Comment (wherka-ama) [reply pkg/cluster/internal/providers/podman/provision.go:260]: @BenTheElder : agreed, this is a separate problem. Should I create a separate issue for that and take this part out from the current PR?

**Comment by aojea** [comment]
/hold

holding waiting for podman developers to disclose more about the future plans of the network https://github.com/kubernetes-sigs/kind/issues/2882#issuecomment-1217734268

we already have several places that fork based on the  podman version, and we can't keep growing like that

File: pkg/cluster/internal/providers/podman/provision.go (hunk 1 above)
Comment (wherka-ama) [reply pkg/cluster/internal/providers/podman/provision.go]: Corrected. Thanks!

File: pkg/cluster/internal/providers/podman/provision.go (hunk 2 above)
Comment (wherka-ama) [reply pkg/cluster/internal/providers/podman/provision.go]: > This should be a separate commit or PR, and it should probably not reimplement naming the control plane endpoint with different logic.

Agreed. I've taken it out and will create a separate issue + another PR as an attempt to tackle the no_proxy inconsistencies. Thanks!

File: pkg/cluster/internal/providers/podman/provision.go [hunk 3]
```diff
@@ -264,15 +265,54 @@ func getProxyEnv(cfg *config.Cluster, networkName string) (map[string]string, er
...
+				Gateway string `json:"gateway"`
+				Subnet  string `json:"subnet"`
+			} `json:"ranges"`
+		} `json:"ipam,omitempty"`
+	} `json:"plugins"`
+}
+
 func getSubnets(networkName string) ([]string, error) {
-	// TODO: unmarshall json and get rid of this complex query
-	format := `{{ range (index (index (index (index . "plugins") 0 ) "ipam" ) "ranges")}}{{ index ( index . 0 ) "subnet" }} {{end}}`
```
Comment (aojea) [inline pkg/cluster/internal/providers/podman/provision.go:269]: I can't remember now exactly, but I think we only care about the first plugin and subnet

Comment (wherka-ama) [reply pkg/cluster/internal/providers/podman/provision.go:269]: > I can't remember now exactly, but I think we only care about the first plugin and subnet

@aojea : would you like me to change that to replicate this behaviour here as well? I'm not sure what was the reasoning behind the first implementation, so would be happy to follow your guidance.

Comment (aojea) [reply pkg/cluster/internal/providers/podman/provision.go:269]: yeah. match the first implementation, it is safer

File: pkg/cluster/internal/providers/podman/provision.go (hunk 2 above)
Comment (BenTheElder) [reply pkg/cluster/internal/providers/podman/provision.go]: thank you!

File: pkg/cluster/internal/providers/podman/provision.go [hunk 4]
```diff
@@ -264,15 +265,54 @@ func getProxyEnv(cfg *config.Cluster, networkName string) (map[string]string, er
...
+			Ranges [][]struct {
+				Gateway string `json:"gateway"`
+				Subnet  string `json:"subnet"`
+			} `json:"ranges"`
+		} `json:"ipam,omitempty"`
+	} `json:"plugins"`
+}
+
 func getSubnets(networkName string) ([]string, error) {
-	// TODO: unmarshall json and get rid of this complex query
```
Comment (BenTheElder) [inline pkg/cluster/internal/providers/podman/provision.go:268]: New TODO: unit test parsing with some known real data to prevent regressions. (doesn't need to be this PR, we should have done that before)

**Review by BenTheElder — COMMENTED** [review]
/lgtm
/approve
thanks so much!

File: pkg/cluster/internal/providers/podman/provision.go (hunk 4 above)
Comment (wherka-ama) [reply pkg/cluster/internal/providers/podman/provision.go:268]: > New TODO: unit test parsing with some known real data to prevent regressions. (doesn't need to be this PR, we should have done that before)

@BenTheElder : funny you mentioned that cause I thought there is something missing in that area indeed and I've already started working on unit tests to cover at least what I've touched ;-)
We can either do it in the next PR, which I'm happy to deliver pretty soon(couple of days max, but hopefully tomorrow) or we can put this one on hold and bundle it with unit tests.  Please do let me know what's your preference and I'll go with it.

Comment (BenTheElder) [reply pkg/cluster/internal/providers/podman/provision.go:268]: another PR when you have time would be great!! no rush, I'm trying to land one or two more fixes ahead of k8s 1.25 / kind v0.15 but I think unit tests can wait / are more forward thinking 😅

File: pkg/cluster/internal/providers/podman/provision.go (hunk 3 above)
Comment (wherka-ama) [reply pkg/cluster/internal/providers/podman/provision.go:269]: @aojea : I have a couple of points before I actually push any further commits to this PR

1. At one point in the issue (https://github.com/kubernetes-sigs/kind/issues/2882#issuecomment-1218019479) you pointed at the existing checks which I could use as my guideline. I'm fine with the version check(if we are to stick with explicit flows), but the part where there was a check to see if we are in the Rootless mode doesn't seem appropriate to me. The trigger of the error here is the presence of the environment variables linked with http(s) proxy. It makes no difference whether we run it as root or not. I've tested it and we have a consistent behaviour here. It is indeed generic in that regard.
2. I did a bit of a refactoring of the code to be very explicit with the versions/layouts used so that allows on very easy decommissioning of the legacy flow when the time comes. However, I must stress it became excessively verbose and I count on your judgement here to decide if that's the direction we should take.
3. In the refactored version I'm referring to the first plugin/subnet - mimicking the old query.  

@BenTheElder: I would welcome your opinion on that aspect as well.

[patch.txt](https://github.com/kubernetes-sigs/kind/files/9365140/patch.txt)
